networkit/cpp/centrality/PageRank.cpp

```diff
@@ -5,73 +5,83 @@
 
 namespace NetworKit {
 
-PageRank::PageRank(const Graph& G, double damp, double tol):
-        Centrality(G, true), damp(damp), tol(tol)
-{
-
-}
+PageRank::PageRank(const Graph &G, double damp, double tol)
+    : Centrality(G, true), damp(damp), tol(tol) {}
 
 void PageRank::run() {
     Aux::SignalHandler handler;
-    count n = G.numberOfNodes();
-    count z = G.upperNodeIdBound();
-    double oneOverN = 1.0 / (double) n;
-    double teleportProb = (1.0 - damp) / (double) n;
-    scoreData.resize(z, oneOverN);
+    const auto n = G.numberOfNodes();
+    const auto z = G.upperNodeIdBound();
+
+    const auto teleportProb = (1.0 - damp) / static_cast<double>(n);
+    scoreData.resize(z, 1.0 / static_cast<double>(n));
     std::vector<double> pr = scoreData;
-    bool isConverged = false;
 
     std::vector<double> deg(z, 0.0);
-    G.parallelForNodes([&](node u) {
-        deg[u] = (double) G.weightedDegree(u);
+    G.parallelForNodes([&](const node u) { deg[u] = static_cast<double>(G.weightedDegree(u)); });
+
+    iterations = 0;
+
+    auto sumL1Norm = [&](const node u) { return std::abs(scoreData[u] - pr[u]); };
+
+    auto sumL2Norm = [&](const node u) {
+        const auto d = scoreData[u] - pr[u];
+        return d * d;
+    };
+
+    auto converged([&]() {
+        if (iterations >= maxIterations) {
+            return true;
+        }
+
+        if (norm == Norm::L2Norm) {
+            return std::sqrt(G.parallelSumForNodes(sumL2Norm)) <= tol;
+        }
+
+        return G.parallelSumForNodes(sumL1Norm) <= tol;
     });
 
-    while (! isConverged) {
+    bool isConverged = false;
+    do {
         handler.assureRunning();
-        G.balancedParallelForNodes([&](node u) {
+        G.balancedParallelForNodes([&](const node u) {
             pr[u] = 0.0;
-            G.forInEdgesOf(u, [&](node u, node v, edgeweight w) {
-                // note: inconsistency in definition in Newman's book (Ch. 7) regarding directed graphs
-                // we follow the verbal description, which requires to sum over the incoming edges
+            G.forInEdgesOf(u, [&](const node u, const node v, const edgeweight w) {
+                // note: inconsistency in definition in Newman's book (Ch. 7) regarding directed
+                // graphs we follow the verbal description, which requires to sum over the incoming
+                // edges
                 pr[u] += scoreData[v] * w / deg[v];
             });
             pr[u] *= damp;
             pr[u] += teleportProb;
         });
 
-        auto converged([&]() {
-            double diff = G.parallelSumForNodes([&](node u) {
-                double d = scoreData[u] - pr[u];
-                return d * d;
-            });
-//			TRACE("sqrt(diff): ", sqrt(diff));
-            return (sqrt(diff) <= tol);
-        });
-
+        ++iterations;
         isConverged = converged();
-        scoreData = pr;
-    }
+        std::copy(pr.begin(), pr.end(), scoreData.begin());
```

Comment:
Why is this a copy and not a swap? As far as I can see the values in `pr` are not read in the next iteration and thus replacing them by `scoreData` should be no problem.

Reply:
That's true, thanks I fixed that.